---
title: "1. Opossum, Split Classes"
output:
---

```{r setup, include=FALSE}
# This sets the project root based on the repo structure.
# If you move this file, you may to set the root manually to find config.R
knitr::opts_knit$set(root.dir = dirname(dirname(rstudioapi::getSourceEditorContext()$path)))
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

```{r libraries}
library(Seurat)
library(SeuratDisk)
library(reticulate)
library(scrubletR)
library(ggplot2)
library(dplyr)
library(data.table)
library(tidyverse)
library(comparatome)
source("config.R")
```

```{r preprocess_data}
# Path to 10x output data and orthology table for gene mapping
data_path <- paste0(dir.list$central, "samples/")
sample_IDs <- c('OpossumV1-1A', 'OpossumV1-1B', 'OpossumV1-2A', 'OpossumV1-2B')
mapping_path <- paste0(dir.list$central, "genome/Opossum_Mouse_GeneMapping_EnsemblBioMart.txt")
# Preprocess (incl. cell/gene filtering)
data <- PreprocessData(data_path, sample_IDs, "Opossum_V1", mapping_path)
# Grab the Seurat object for direct analysis
obj.opossum <- data$obj
obj.opossum$species <- "Opossum"
```

```{r qc_plots, fig.height=3, fig.width=4}
# Plot nFeature and nCount quality control
PlotQC(data)
```

```{r initial_clustering}
# Initial clustering and diagnostic plots
obj.opossum <- ClusterWithSCT(obj.opossum, 1)
clust.plots <- PlotClusters(obj.opossum)
# Save plots for Figure S1
save.plots <- list("dimplot_cluster" = "S1E_Opossum-All-Cluster-DimPlot",
                   "dimplot_sample" = "S1F_Opossum-All-Sample-DimPlot",
                   "barplot_sample" = "S1G_Opossum-All-Sample-BarPlot",
                   "barplot_doublet" = "S1H_Opossum-All-Doublet-BarPlot")
for (sp in names(save.plots)) {
  SavePNGandSVG(clust.plots[[sp]], dir.list$figS1$plots, save.plots[[sp]])
}
```

```{r filter_clusters}
# Remove high-doublet clusters and re-cluster (skip for this data)
obj.opossum <- subset(obj.opossum, idents = c(NA), invert = TRUE)
obj.opossum <- ClusterWithSCT(obj.opossum, 1)
PlotClusters(obj.opossum)
```

```{r canonical_markers, fig.width=12, fig.height=6}
# Examine canonical subclass marker gene expression
canon.markers <- list(
                    class = c("Snap25", "Sv2b", "Gad1", "Gad2"),
                    L23 = c("Cux2", "Ccbe1", "Mdga1", "Stard8"),
                    L4 = c("Whrn", "Rorb"),
                    L5IT = c("Bcl11b"),
                    L5NP = c("Nxph1", "Tshz2", "Trhr", "Slc17a8"),
                    L5PT = c("Bcl6", "Erg", "Reln"),
                    L6CT = c("Foxp2", "Syt6"),
                    L6IT = c("Zfp804b", "Cdh9"),
                    L6b = c("Ctgf", "Inpp4b", "Svil"),
                    Pvalb = c("Pvalb", "Myo5b"), 
                    Sst = c("Sst", "Chodl"),
                    Vip = c("Vip"),
                    Lamp5 = c("Lamp5", "Sv2c"),
                    Other = c("Frem1", "Stac", "Sncg", "Meis2"),
                    Astro = c("Aldh1l1", "Grin2c", "Phkg1"),
                    Endo = c("Pecam1", "Mecom", "Flt1"),
                    Micro = c("Cx3cr1", "Arhgap45", "Inpp5d"),
                    OD = c("Enpp6", "Mog", "St18"),
                    OPC = c("Pdgfra", "Cacng4"),
                    VLMC = c("Slc6a13", "Slc47a1", "Bnc2")
                      )
DefaultAssay(obj.opossum) <- "RNA"
DotPlot(obj.opossum, features = canon.markers, cols = c("lightgrey", "red"), scale = FALSE) + 
                         theme(axis.text.x = element_text(angle = 90, vjust = 0.25)) + 
                         theme(panel.background = element_rect(fill = "white", color = NA), plot.background = element_rect(fill = "white", color = NA)) + NoLegend()
```

```{r class_marker_dotplot, fig.width=1.5, fig.height=6}
# Plot canonical class markers for Figure S1
DefaultAssay(obj.opossum) <- "RNA"
p <- DotPlot(obj.opossum, features = c("Snap25", "Sv2b", "Gad1", "Gad2"), 
                          cols = c("lightgrey", "red"), scale = FALSE) + 
     theme(axis.text.x = element_text(angle = 90, vjust = 0.25)) + 
     theme(panel.background = element_rect(fill = "white", color = NA), 
           plot.background = element_rect(fill = "white", color = NA)) + 
     NoLegend()
print(p)
SavePNGandSVG(p, dir.list$figS1$plots, "S1I_Opossum-All-Class-DotPlot")
```

```{r label_classes}
# Label classes in full Seurat object
# Glutamatergic
cells.glutamatergic <- WhichCells(obj.opossum, ident = c(1, 4, 5, 8, 9, 10, 11, 13, 14, 15, 17, 20, 24))
obj.opossum$class <- ifelse(colnames(obj.opossum) %in% cells.glutamatergic, "glutamatergic", "ambiguous")
# GABAergic
cells.gabaergic <- WhichCells(obj.opossum, ident = c(6, 16, 18, 21, 22, 23, 25))
obj.opossum$class[cells.gabaergic] <- "gabaergic"
# Non-neuronal
cells.nonneuronal <- WhichCells(obj.opossum, ident = c(2, 3, 7, 12, 19, 26, 27))
obj.opossum$class[cells.nonneuronal] <- "nonneuronal"
```

```{r plot_classes}
# Plot classes in all-cell UMAP
p <- DimPlot(obj.opossum, reduction = "umap", group.by = "class", label = FALSE, raster = FALSE, shuffle = TRUE) + xlim(-18, 18) + ylim(-18, 18) + coord_equal()
print(p)
SavePNGandSVG(p, dir.list$figS1$plots, "S1J_Opossum-All-Class-DimPlot")
```

```{r save_split_classes}
# Save the original clustering result for future reference
obj.opossum$all_SCT_snn_res.1 <- obj.opossum$SCT_snn_res.1

# Save classes into separate object for downstream processing
Idents(obj.opossum) <- "class"
# Glutamatergic
obj.opossum.glutamatergic <- subset(obj.opossum, idents = "glutamatergic")
saveRDS(obj.opossum.glutamatergic, paste0(dir.list$snrna$seurat$raw, "opossum_v1_glutamatergic_raw.rds"))
# GABAergic
obj.opossum.gabaergic <- subset(obj.opossum, idents = "gabaergic")
saveRDS(obj.opossum.gabaergic, paste0(dir.list$snrna$seurat$raw, "opossum_v1_gabaergic_raw.rds"))
# Non-neuronal
obj.opossum.nonneuronal <- subset(obj.opossum, idents = "nonneuronal")
saveRDS(obj.opossum.nonneuronal, paste0(dir.list$snrna$seurat$raw, "opossum_v1_nonneuronal_raw.rds"))
```
